documentation/design-idna.md

```diff
@@ -0,0 +1,244 @@
+#5: ± MODIFY foo.рф.com (10 рф.com. ttl=14400) -> (10 foo.рф.com. ttl=14400)
+#6: ± MODIFY foo.xn--p1ai.com (10 xn--p1ai.com. ttl=14400) -> (10 foo.xn--p1ai.com. ttl=14400)
+#7: + CREATE foo1.example.com MX 10 mxfoo.example.com (ttl=14400)
+#8: ± MODIFY foo2.example.com (10 example.com. ttl=14400) -> (10 foo.example.com. ttl=14400)
+#9: + CREATE foo3.example.com MX 10 xn--p1ai.com. ⟬рф.com.⟭ (ttl=14400)
+#10: ± MODIFY foo4.example.com (10 xn--p1ai.com. ⟬рф.com.⟭ ttl=14400) -> (10 foo.example.com. ttl=14400)
+#11: ± MODIFY foo5.example.com (10 example.com. ttl=14400) -> (10 xn--p1ai.com. ⟬рф.com.⟭ ttl=14400)
+```
+
+Now here are the same examples with `()` changed to `❮❯`:
+
+```
+#1: + CREATE foo.рф.com ❮foo.xn--p1ai.com❯ MX 10 рф.com. ❮xn--p1ai.com.❯ (ttl=14400)
+#2: + CREATE foo.xn--p1ai.com ❮foo.рф.com❯ MX 10 xn--p1ai.com. ❮рф.com.❯ (ttl=14400)
+#3: ± MODIFY foo.xn--p1ai.com ❮foo.рф.com❯ (10 xn--p1ai.com. ❮рф.com.❯ ttl=14400) -> (10 foo.xn--p1ai.com. ❮foo.рф.com.❯ ttl=14400)
+#4: ± MODIFY foo.рф.com ❮foo.xn--p1ai.com❯ (10 рф.com. ❮xn--p1ai.com.❯ ttl=14400) -> (10 foo.рф.com. ❮foo.xn--p1ai.com.❯ ttl=14400)
+#5: ± MODIFY foo.рф.com (10 рф.com. ttl=14400) -> (10 foo.рф.com. ttl=14400)
+#6: ± MODIFY foo.xn--p1ai.com (10 xn--p1ai.com. ttl=14400) -> (10 foo.xn--p1ai.com. ttl=14400)
+#7: + CREATE foo1.example.com MX 10 mxfoo.example.com (ttl=14400)
+#8: ± MODIFY foo2.example.com (10 example.com. ttl=14400) -> (10 foo.example.com. ttl=14400)
+#9: + CREATE foo3.example.com MX 10 xn--p1ai.com. ❮рф.com.❯ (ttl=14400)
+#10: ± MODIFY foo4.example.com (10 xn--p1ai.com. ❮рф.com.❯ ttl=14400) -> (10 foo.example.com. ttl=14400)
+#11: ± MODIFY foo5.example.com (10 example.com. ttl=14400) -> (10 xn--p1ai.com. ❮рф.com.❯ ttl=14400)
+```
+
+
+# Design
+
+In general, DNSControl will store domains and labels in multiple
+formats: (1) in the original
+format the user specified (downcased), in ASCII, and in Unicode, and
+in a format useful for displaying to users.  This way
+providers do not have to do conversions.
+
+When the user used Unicode:
+
+* Original: рф.com
+* ASCII: xn--p1ai.com
+* Unicode: рф.com
+* Display: xn--p1ai.com (рф.com)
+
+When the user used ASCII:
+
+* Original: xn--p1ai.com
+* ASCII: xn--p1ai.com
+* Unicode: рф.com
+* Display: xn--p1ai.com (рф.com)
+
+NOTE: User input is downcased.  If the user input is `D('xn--P1AI.COM')` the Original field would be `xn--p1ai.com` and so on.
+
+Memory usage will be minimized by using Go's slices.  In the above
+example, the Display string would be generated first, the others would
+be slices of that string.
+
+
+```
+models.DomainConfig:
+
+   .Name: the name from D() after downcased via unicode.ToLower()
+   .NameASCII: The name stored after calling ToASCII() (with ACE prefix if any Unicode chars are present)
+   .NameUnicode: The name stored after calling ToUnicode()
+   .NameDisplay: if .NameASCII != .NameUnicode, store as "ascii (unicode)"
```

Comment:
This is confusing. Shouldn't this be the bigger slice that `Name` and `IDN` would be subslices of?